frame/contracts/src/lib.rs

```diff
@@ -967,6 +989,156 @@ struct InternalOutput<T: Config, O> {
 	result: Result<O, ExecError>,
 }
 
+/// Helper trait to wrap contract execution entry points into a signle function
+/// [`Invokable::run_guarded`].
+trait Invokable<T: Config, O> {
+	/// Single entry point to contract execution.
+	/// Downstream execution flow is branched by implementations of [`Invokable`] trait:
+	///
+	/// - [`InstantiateInput::run`] runs contract instantiation,
+	/// - [`CallInput::run`] runs contract call.
+	///
+	/// We enforce a re-entrancy guard here by initializing and checking a boolean flag through a
+	/// global reference.
+	fn run_guarded(&self, context: ExecContext<T>) -> InternalOutput<T, O> {
+		let gas_limit = context.gas_limit;
+		executing_contract::using_once(&mut false, || {
+			executing_contract::with(|f| {
+				// Fail if already entered contract execution
+				if *f {
+					return Err(())
+				}
+				// We are entering contract execution
+				*f = true;
+				Ok(())
+			})
+			.unwrap_or(Ok(())) // Should always be `Some`, still we make it safe.
```

Comment:
It might be somewhat safer but a lot more confusing to not panic. There is a simple proof:

```suggestion
			.expect("Returns `Ok` if called within `using once`. Is is syntactically obvious that this is the case; qed")
```